BUG: In Login activity, app crashes during orientation change.

**Summary:** 

When orientation change occurs in login activity, it crashes the app.

**Steps to reproduce:** 

How can we reproduce the issue?
1. Go the login screen (Logout if already logged in).
2. Change the phone orientation.

**Add System logs:**
```
02-26 07:40:20.076 19163-19163/fr.free.nrw.commons.debug E/AndroidRuntime: FATAL EXCEPTION: main
                                                                           Process: fr.free.nrw.commons.debug, PID: 19163
                                                                           java.lang.RuntimeException: Unable to start activity ComponentInfo{fr.free.nrw.commons.debug/fr.free.nrw.commons.auth.LoginActivity}: java.lang.IllegalStateException: Required view 'login_credentials' with ID 2131689634 for field 'loginCredentials' was not found. If this view is optional add '@Nullable' (fields) or '@Optional' (methods) annotation.
                                                                               at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2696)
                                                                               at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2757)
                                                                               at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4553)
                                                                               at android.app.ActivityThread.-wrap19(ActivityThread.java)
                                                                               at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1512)
                                                                               at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                               at android.os.Looper.loop(Looper.java:163)
                                                                               at android.app.ActivityThread.main(ActivityThread.java:6237)
                                                                               at java.lang.reflect.Method.invoke(Native Method)
                                                                               at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:877)
                                                                               at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
                                                                            Caused by: java.lang.IllegalStateException: Required view 'login_credentials' with ID 2131689634 for field 'loginCredentials' was not found. If this view is optional add '@Nullable' (fields) or '@Optional' (methods) annotation.
                                                                               at butterknife.internal.Utils.findRequiredView(Utils.java:92)
                                                                               at butterknife.internal.Utils.findRequiredViewAsType(Utils.java:104)
                                                                               at fr.free.nrw.commons.auth.LoginActivity_ViewBinding.<init>(LoginActivity_ViewBinding.java:37)
                                                                               at java.lang.reflect.Constructor.newInstance0(Native Method)
                                                                               at java.lang.reflect.Constructor.newInstance(Constructor.java:430)
                                                                               at butterknife.ButterKnife.createBinding(ButterKnife.java:199)
                                                                               at butterknife.ButterKnife.bind(ButterKnife.java:124)
                                                                               at fr.free.nrw.commons.auth.LoginActivity.onCreate(LoginActivity.java:95)
                                                                               at android.app.Activity.performCreate(Activity.java:6847)
                                                                               at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1119)
                                                                               at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2649)
                                                                               at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2757) 
                                                                               at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4553) 
                                                                               at android.app.ActivityThread.-wrap19(ActivityThread.java) 
                                                                               at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1512) 
                                                                               at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                               at android.os.Looper.loop(Looper.java:163) 
                                                                               at android.app.ActivityThread.main(ActivityThread.java:6237) 
                                                                               at java.lang.reflect.Method.invoke(Native Method) 
                                                                               at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:877) 
                                                                               at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767) 
```

**Device and Android version:** 

Android 7 MIUI V9.0
 
 **Commons app version:** 

2.6.7-debug-master~e9602f5


**Screen-shots:** 

![crashgif](https://user-images.githubusercontent.com/20313518/36650406-3f1e91e8-1ac9-11e8-837d-bdc5585ca89e.gif)

**I am working on the issue**

Can anybody confirm the bug?

**Update**

TextView
                    android:id="@+id/login_credentials"
                    android:layout_width="match_parent"
                    android:layout_height="wrap_content"
                    android:layout_below="@id/title"
                    android:layout_marginEnd="@dimen/standard_gap"
                    android:layout_marginLeft="@dimen/standard_gap"
                    android:layout_marginRight="@dimen/standard_gap"
                    android:layout_marginStart="@dimen/standard_gap"
                    android:paddingBottom="@dimen/small_gap"
                    android:paddingTop="@dimen/small_gap"
                    android:textAlignment="center"
                    android:textColor="@color/secondaryDarkColor"
                    tools:text="@string/login_credential" 

This TextView is only present in portrait mode of activity_login.xml
Added this to landscape and x-large activity_login.xml solves the bug.
PR #1208 fixes this issue.

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -49,7 +49,7 @@ dependencies {
     compile 'com.facebook.stetho:stetho:1.5.0'
 
     testCompile 'junit:junit:4.12'
-    testCompile 'org.robolectric:robolectric:3.4'
+    testCompile 'org.robolectric:robolectric:3.7.1'
 
     testCompile 'com.squareup.okhttp3:mockwebserver:3.8.1'
     androidTestCompile 'com.squareup.okhttp3:mockwebserver:3.8.1'
@@ -78,7 +78,7 @@ dependencies {
     androidTestImplementation "org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
 
     testImplementation 'junit:junit:4.12'
-    testImplementation 'org.robolectric:robolectric:3.4'
+    testImplementation 'org.robolectric:robolectric:3.7.1'
     testImplementation 'org.mockito:mockito-all:1.10.19'
 
     testImplementation 'com.squareup.okhttp3:mockwebserver:3.8.1'

--- a/app/src/main/res/layout-land/activity_login.xml
+++ b/app/src/main/res/layout-land/activity_login.xml
@@ -13,12 +13,12 @@
         <android.support.v7.widget.CardView
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
-            android:layout_marginTop="@dimen/large_gap"
             android:layout_marginBottom="@dimen/standard_gap"
             android:layout_marginEnd="@dimen/standard_gap"
             android:layout_marginLeft="@dimen/standard_gap"
             android:layout_marginRight="@dimen/standard_gap"
             android:layout_marginStart="@dimen/standard_gap"
+            android:layout_marginTop="@dimen/large_gap"
             app:cardCornerRadius="4dp"
             app:cardElevation="4dp">
 
@@ -40,11 +40,26 @@
                     android:textColor="@android:color/white"
                     android:textSize="@dimen/heading_text_size" />
 
+                <TextView
+                    android:id="@+id/login_credentials"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_below="@id/title"
+                    android:layout_marginEnd="@dimen/standard_gap"
+                    android:layout_marginLeft="@dimen/standard_gap"
+                    android:layout_marginRight="@dimen/standard_gap"
+                    android:layout_marginStart="@dimen/standard_gap"
+                    android:paddingBottom="@dimen/standard_gap"
+                    android:paddingTop="@dimen/small_gap"
+                    android:textAlignment="center"
+                    android:textColor="@color/secondaryDarkColor"
+                    tools:text="@string/login_credential" />
+
                 <FrameLayout
                     android:id="@+id/error_message_container"
                     android:layout_width="match_parent"
                     android:layout_height="wrap_content"
-                    android:layout_below="@id/title"
+                    android:layout_below="@id/login_credentials"
                     android:visibility="gone"
                     tools:visibility="visible">
 
@@ -56,6 +71,7 @@
                         android:layout_marginLeft="@dimen/standard_gap"
                         android:layout_marginRight="@dimen/standard_gap"
                         android:layout_marginStart="@dimen/standard_gap"
+                        android:gravity="center"
                         android:paddingBottom="@dimen/small_gap"
                         android:paddingTop="@dimen/small_gap"
                         android:textColor="@color/secondaryDarkColor"
@@ -149,8 +165,8 @@
 
                     <Button
                         android:id="@+id/signupButton"
-                        android:layout_width="0dp"
                         style="@style/Widget.AppCompat.Button.Borderless.Colored"
+                        android:layout_width="0dp"
                         android:layout_height="wrap_content"
                         android:layout_gravity="center_horizontal"
                         android:layout_marginEnd="@dimen/small_gap"
@@ -160,8 +176,8 @@
 
                     <Button
                         android:id="@+id/loginButton"
-                        android:layout_width="0dp"
                         style="@style/Widget.AppCompat.Button.Colored"
+                        android:layout_width="0dp"
                         android:layout_height="wrap_content"
                         android:layout_marginLeft="@dimen/small_gap"
                         android:layout_marginStart="@dimen/small_gap"
@@ -174,13 +190,13 @@
                 <fr.free.nrw.commons.ui.widget.HtmlTextView
                     android:id="@+id/about_privacy_policy"
                     style="?android:textAppearanceSmall"
-                    android:layout_below="@id/buttonFrame"
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
-                    android:layout_marginBottom="@dimen/standard_gap"
+                    android:layout_below="@id/buttonFrame"
+                    android:layout_centerHorizontal="true"
                     android:layout_gravity="center_horizontal"
-                    android:text="@string/about_privacy_policy"
-                    android:layout_centerHorizontal="true"/>
+                    android:layout_marginBottom="@dimen/standard_gap"
+                    android:text="@string/about_privacy_policy" />
 
             </RelativeLayout>
         </android.support.v7.widget.CardView>
@@ -188,19 +204,19 @@
         <android.support.v7.widget.AppCompatImageView
             android:layout_width="64dp"
             android:layout_height="64dp"
-            android:elevation="8dp"
-            tools:ignore="UnusedAttribute"
             android:layout_gravity="center_horizontal"
-            app:srcCompat="@drawable/blue_rinse_circle" />
+            android:elevation="8dp"
+            app:srcCompat="@drawable/blue_rinse_circle"
+            tools:ignore="UnusedAttribute" />
 
         <android.support.v7.widget.AppCompatImageView
             android:layout_width="42dp"
-            tools:ignore="UnusedAttribute"
             android:layout_height="42dp"
             android:layout_gravity="center_horizontal"
             android:layout_marginTop="8dp"
             android:elevation="8dp"
-            app:srcCompat="@drawable/commons_logo_large" />
+            app:srcCompat="@drawable/commons_logo_large"
+            tools:ignore="UnusedAttribute" />
 
     </FrameLayout>
 

--- a/app/src/main/res/layout-xlarge/activity_login.xml
+++ b/app/src/main/res/layout-xlarge/activity_login.xml
@@ -13,12 +13,12 @@
         <android.support.v7.widget.CardView
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
-            android:layout_marginTop="@dimen/large_gap"
             android:layout_marginBottom="@dimen/standard_gap"
             android:layout_marginEnd="@dimen/standard_gap"
             android:layout_marginLeft="@dimen/standard_gap"
             android:layout_marginRight="@dimen/standard_gap"
             android:layout_marginStart="@dimen/standard_gap"
+            android:layout_marginTop="@dimen/large_gap"
             app:cardCornerRadius="4dp"
             app:cardElevation="4dp">
 
@@ -40,6 +40,21 @@
                     android:textColor="@android:color/white"
                     android:textSize="@dimen/heading_text_size" />
 
+                <TextView
+                    android:id="@+id/login_credentials"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_below="@id/title"
+                    android:layout_marginEnd="@dimen/standard_gap"
+                    android:layout_marginLeft="@dimen/standard_gap"
+                    android:layout_marginRight="@dimen/standard_gap"
+                    android:layout_marginStart="@dimen/standard_gap"
+                    android:paddingBottom="@dimen/small_gap"
+                    android:paddingTop="@dimen/small_gap"
+                    android:textAlignment="center"
+                    android:textColor="@color/secondaryDarkColor"
+                    tools:text="@string/login_credential" />
+
                 <FrameLayout
                     android:id="@+id/error_message_container"
                     android:layout_width="match_parent"
@@ -56,6 +71,7 @@
                         android:layout_marginLeft="@dimen/standard_gap"
                         android:layout_marginRight="@dimen/standard_gap"
                         android:layout_marginStart="@dimen/standard_gap"
+                        android:gravity="center"
                         android:paddingBottom="@dimen/small_gap"
                         android:paddingTop="@dimen/small_gap"
                         android:textColor="@color/secondaryDarkColor"
@@ -149,8 +165,8 @@
 
                     <Button
                         android:id="@+id/signupButton"
-                        android:layout_width="0dp"
                         style="@style/Widget.AppCompat.Button.Borderless.Colored"
+                        android:layout_width="0dp"
                         android:layout_height="wrap_content"
                         android:layout_gravity="center_horizontal"
                         android:layout_marginEnd="@dimen/small_gap"
@@ -160,8 +176,8 @@
 
                     <Button
                         android:id="@+id/loginButton"
-                        android:layout_width="0dp"
                         style="@style/Widget.AppCompat.Button.Colored"
+                        android:layout_width="0dp"
                         android:layout_height="wrap_content"
                         android:layout_marginLeft="@dimen/small_gap"
                         android:layout_marginStart="@dimen/small_gap"
@@ -174,13 +190,13 @@
                 <fr.free.nrw.commons.ui.widget.HtmlTextView
                     android:id="@+id/about_privacy_policy"
                     style="?android:textAppearanceSmall"
-                    android:layout_below="@id/buttonFrame"
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
-                    android:layout_marginBottom="@dimen/standard_gap"
+                    android:layout_below="@id/buttonFrame"
+                    android:layout_centerHorizontal="true"
                     android:layout_gravity="center_horizontal"
-                    android:text="@string/about_privacy_policy"
-                    android:layout_centerHorizontal="true"/>
+                    android:layout_marginBottom="@dimen/standard_gap"
+                    android:text="@string/about_privacy_policy" />
 
             </RelativeLayout>
         </android.support.v7.widget.CardView>
@@ -188,8 +204,8 @@
         <android.support.v7.widget.AppCompatImageView
             android:layout_width="64dp"
             android:layout_height="64dp"
-            android:elevation="8dp"
             android:layout_gravity="center_horizontal"
+            android:elevation="8dp"
             app:srcCompat="@drawable/blue_rinse_circle"
             tools:ignore="UnusedAttribute" />
 
@@ -198,9 +214,9 @@
             android:layout_height="42dp"
             android:layout_gravity="center_horizontal"
             android:layout_marginTop="8dp"
-            tools:ignore="UnusedAttribute"
             android:elevation="8dp"
-            app:srcCompat="@drawable/commons_logo_large" />
+            app:srcCompat="@drawable/commons_logo_large"
+            tools:ignore="UnusedAttribute" />
 
     </FrameLayout>
 

--- a/app/src/main/res/layout/activity_login.xml
+++ b/app/src/main/res/layout/activity_login.xml
@@ -44,20 +44,20 @@
                     android:textColor="@android:color/white"
                     android:textSize="@dimen/heading_text_size" />
 
-                    <TextView
-                        android:id="@+id/login_credentials"
-                        android:layout_width="match_parent"
-                        android:layout_height="wrap_content"
-                        android:layout_below="@id/title"
-                        android:layout_marginEnd="@dimen/standard_gap"
-                        android:layout_marginLeft="@dimen/standard_gap"
-                        android:layout_marginRight="@dimen/standard_gap"
-                        android:layout_marginStart="@dimen/standard_gap"
-                        android:paddingBottom="@dimen/small_gap"
-                        android:paddingTop="@dimen/small_gap"
-                        android:textAlignment="center"
-                        android:textColor="@color/secondaryDarkColor"
-                        tools:text="@string/login_credential"/>
+                <TextView
+                    android:id="@+id/login_credentials"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_below="@id/title"
+                    android:layout_marginEnd="@dimen/standard_gap"
+                    android:layout_marginLeft="@dimen/standard_gap"
+                    android:layout_marginRight="@dimen/standard_gap"
+                    android:layout_marginStart="@dimen/standard_gap"
+                    android:paddingBottom="@dimen/small_gap"
+                    android:paddingTop="@dimen/small_gap"
+                    android:textAlignment="center"
+                    android:textColor="@color/secondaryDarkColor"
+                    tools:text="@string/login_credential" />
 
 
                 <FrameLayout
@@ -67,6 +67,7 @@
                     android:layout_below="@id/login_credentials"
                     android:visibility="gone"
                     tools:visibility="visible">
+
                     <TextView
                         android:id="@+id/error_message"
                         android:layout_width="match_parent"
@@ -75,6 +76,7 @@
                         android:layout_marginLeft="@dimen/standard_gap"
                         android:layout_marginRight="@dimen/standard_gap"
                         android:layout_marginStart="@dimen/standard_gap"
+                        android:gravity="center"
                         android:paddingBottom="@dimen/small_gap"
                         android:paddingTop="@dimen/small_gap"
                         android:textColor="@color/secondaryDarkColor"

